Maven plugin for Helidon Service Inject

tomas-langer opened the pull request:
Maven plugin that:
- generates application binding class (bindings for injection points)
- generates main class (registration of descriptor instances to `InjectConfig` that disables bootstrap service discovery)
- updated tests

To see what happens, build the project with `-Ptests`, and check directory:
`service/tests/inject/toolbox/target/generated-sources/annotations/io/helidon/service/tests/inject/toolbox`

See `ApplicationMain` and `Injection__Binding` classes.

tomas-langer committed:
Small fixes and pre-requisite changes.

tomas-langer committed:
Support for injection Main class

tomas-langer committed:
Maven plugin for Helidon Service Inject

tomas-langer committed:
Tests to validate previous commits.

tomas-langer commented on service/inject/inject/src/main/java/io/helidon/service/inject/InjectionMain.java:
```diff
@@ -0,0 +1,255 @@
+        return false;
+    }
+
+    /**
+     * Called before service descriptors are configured in the generated main class.
+     * This method is invoked from generated {@link #start(String[])}.
+     *
+     * @param configBuilder in-progress config builder
+     */
+    protected void beforeServiceDescriptors(InjectConfig.Builder configBuilder) {
+    }
+
+    /**
+     * Called after service descriptors are configured in the generated main class, before the registry is initialized.
+     * This method is invoked from generated {@link #start(String[])}
+     *
+     * @param configBuilder in-progress config builder
+     */
+    protected void afterServiceDescriptors(InjectConfig.Builder configBuilder) {
+    }
```
Because you are not required to implement them. The only one that is code generated is `serviceDescriptors`, all other methods are for extensibility by the user.

tomas-langer commented on service/inject/maven-plugin/src/main/java/io/helidon/service/inject/maven/plugin/BindingGenerator.java:
```diff
@@ -0,0 +1,584 @@
+        Set<Qualifier> qualifiers = dependencyTo.qualifiers();
+        if (self.contracts().containsAll(dependencyTo.contracts()) && self.qualifiers().equals(qualifiers)) {
+            // criteria must have a single contract for each injection point
+            // if this service implements the contracts actually required, we must look for services with lower weight
```
This is related to injection points - if an injection point requires a contract that the current service implements, we must look for lower weight, as otherwise the current service would be selected to satisfy the injection point of itself.

tomas-langer replied:
I have updated the comment with the above statement.

tomas-langer committed:
Review fixes.

tomas-langer committed:
Rebased on main and updated error message for record field injection.

tomas-langer commented on service/inject/maven-plugin/src/main/java/io/helidon/service/inject/maven/plugin/MavenFiler.java:
```diff
@@ -0,0 +1,155 @@
+
+    @Override
+    public Path writeSourceFile(ClassModel classModel, Object... originatingElements) {
+        TypeName typeName = classModel.typeName();
```
Nice catch.

tomas-langer committed:
Review comments.

tomas-langer committed:
Maven plugin integration tests.
Moved version management for maven plugin to project root pom, moved plugin configuration there as well
A fix in `AptTypeInfoFactory` where a NPE was possible.

tomas-langer commented on service/inject/maven-plugin/pom.xml:
```diff
@@ -0,0 +1,203 @@
+        </dependency>
+    </dependencies>
+
+    <build>
```
Added and testing:
1. defaults
2. custom package, class name for main class, class name for binding
3. disabled generation of main and binding

tomas-langer commented on service/inject/maven-plugin/pom.xml:
```diff
@@ -0,0 +1,203 @@
+                    <version>${version.plugin.plugin}</version>
+                    <configuration>
+                        <mojoDependencies>
+                            <!-- we need this since 19ea is not really supported -->
```
Removed

tomas-langer commented on service/inject/maven-plugin/pom.xml:
```diff
@@ -0,0 +1,203 @@
+            <version>${version.plugin.api}</version>
+            <scope>provided</scope>
+            <exclusions>
+                <exclusion>
```
Removed.

tomas-langer commented on service/inject/maven-plugin/pom.xml:
```diff
@@ -0,0 +1,203 @@
+            <version>${version.plugin.project}</version>
+            <scope>provided</scope>
+            <exclusions>
+                <exclusion>
```
Removed.

tomas-langer commented on service/inject/maven-plugin/pom.xml:
```diff
@@ -0,0 +1,203 @@
+        </dependency>
+        <dependency>
+            <!--
+            we cannot use helidon-service-inject as the SPI is required on classpath of application,
```
Removed the comment.
Conflicts are coming from the classloader (as we would load classes the registry classes in a different classloader than the application classes (maven vs. application), which resulted in some classes not being seen.

romain-grecourt replied:
Perhaps the plugin should setup a class-loader using the project dependencies to micmic the actual application ?

tomas-langer commented on bom/pom.xml:
```diff
@@ -1506,6 +1506,11 @@
                 <artifactId>helidon-service-inject</artifactId>
                 <version>${helidon.version}</version>
             </dependency>
+            <dependency>
```
I would prefer to leave the application updates to Helidon declarative, as this requires actual applications using it to verify it is done correctly, and that seemed like a bit bigger task.

romain-grecourt replied:
This plugin will be usable in 4.2.0, we should provide proper version management for it.

tomas-langer committed:
Use `release` if available for codegen compiler
Switch hint log level to trace in maven plugin
Removed bad info log, added useful info log to maven plugin

tomas-langer committed:
Fix qualifiers

tomas-langer committed:
Updates to registry (less dependencies, single module)

tomas-langer committed:
Single registry.
Cleaned up dependencies.
Support for registry in builder.

tomas-langer committed:
Fixed spotbugs, checkstyle etc.

tomas-langer committed:
Copyright fixes.

tomas-langer committed:
New main class generator, binding generator. All tests passing in services

tomas-langer committed:
Fix config provider to honor default config sources when meta configuration is not present.
Added a test to validate it.

tomas-langer committed:
Updated docs.

tomas-langer committed:
A small update to the generated code and registry, to register shutdown handler for started registry managers.

tomas-langer committed:
Update documentation

tomas-langer committed:
Fix versions of plugins and release of java in maven invoker tests

tomas-langer committed:
Ensure bindings must not fail if no descriptor discovered, as there may be a service that requires a dependency, but itself is not required by any injection point.

tomas-langer committed:
Fixed service supply. Now we try to use all matchin service descriptors even for a single instance, as any may return an empty optional (if appropriate supplier type), and we would not yield an instance, even when we should.
As a result the method "explodeFilterAndSort" was going through all service manager and collected all instances - which resulted in errors if a provider with lower weight was not capable of creating an intance because of unsatisfied dependency.

tomas-langer committed:
Plugin management for services Maven plugin

tomas-langer committed:
Fix accidental replacement of Injection -> Service
